File: ui/src/api/endpoints/equipment.ts
/**
 * Equipment API Endpoints
 */
import apiClient from '../client';

export interface EquipmentMaster {
  id: string;
  tenant_id: string;
  equipment_code: string;
  equipment_name: string;
  equipment_type: string;
  equipment_category?: string;
  line_code: string;
  position_in_line?: number;
  manufacturer?: string;
  model?: string;
  serial_no?: string;
  firmware_version?: string;
  install_date?: string;
  warranty_end_date?: string;
  standard_cycle_time_sec?: number;
  max_capacity_per_hour?: number;
  ip_address?: string;
  port?: number;
  communication_protocol?: string;
  last_pm_date?: string;
  next_pm_date?: string;
  pm_interval_days: number;
  status: string;
  attributes?: Record<string, any>;
  created_at: string;
  updated_at?: string;
}

export interface EquipmentStatus {
  id: string;
  tenant_id: string;
  equipment_code: string;
  status_timestamp: string;
  status: 'running' | 'idle' | 'setup' | 'alarm' | 'breakdown' | 'maintenance' | 'offline' | 'waiting';
  previous_status?: string;
  status_reason?: string;
  alarm_code?: string;
  alarm_message?: string;
  alarm_severity?: 'info' | 'warning' | 'error' | 'critical' | 'fatal';
  production_order_no?: string;
  product_code?: string;
  operator_code?: string;
  speed_actual?: number;
  speed_target?: number;
  temperature_actual?: number;
  temperature_target?: number;
  counters?: Record<string, any>;
  parameters?: Record<string, any>;
  created_at: string;
}

export interface EquipmentCurrentStatus {
  equipment_code: string;
  equipment_name: string;
  equipment_type: string;
  line_code: string;
  position_in_line?: number;
  status: string;
  status_since?: string;
  current_order_no?: string;
  product_code?: string;
  alarm_code?: string;
  alarm_message?: string;
}

export interface EquipmentOEE {
  id: string;
  tenant_id: string;
  calculation_date: string;
  shift_code?: string;
  equipment_code: string;
  line_code: string;
  planned_time_min: number;
  actual_run_time_min: number;
  downtime_min: number;
  setup_time_min: number;
  idle_time_min: number;
  ideal_cycle_time_sec?: number;
  actual_cycle_time_sec?: number;
  total_count: number;
  good_count: number;
  defect_count: number;
  oee?: number;
  availability?: number;
  performance?: number;
  quality?: number;
  downtime_breakdown?: Record<string, any>;
  defect_breakdown?: Record<string, any>;
  calculated_at?: string;
  created_at: string;
}

export interface OEETrend {
  date: string;
  oee: number;
  availability: number;
  performance: number;
  quality: number;
}

export interface DowntimeEvent {
  id: string;
  tenant_id: string;
  event_no: string;
  equipment_code: string;
  line_code: string;
  start_time: string;
  end_time?: string;
  duration_min?: number;
  downtime_category: string;
  downtime_code: string;
  downtime_reason?: string;
  alarm_code?: string;
  alarm_message?: string;
  production_order_no?: string;
  product_code?: string;
  operator_code?: string;
  root_cause?: string;
  corrective_action?: string;
  maintenance_ticket_no?: string;
  impact_qty?: number;
  impact_cost?: number;
  status: 'open' | 'in_progress' | 'resolved' | 'closed';
  resolved_by?: string;
  resolved_at?: string;
  notes?: string;
  reported_by?: string;
  created_at: string;
  updated_at?: string;
}

export interface LineStatus {
  line_code: string;
  line_name: string;
  line_type: string;
  status: string;
  equipment_count: number;
  running_count: number;
  idle_count: number;
  down_count: number;
  current_oee?: number;
  today_production?: number;
  today_defects?: number;
}

// API Functions

export const getEquipmentList = async (params?: {
  line_code?: string;
  equipment_type?: string;
  status?: string;
}): Promise<EquipmentMaster[]> => {
  const response = await apiClient.get('/mes/equipment', { params });
  return response.data;
};

export const getEquipment = async (equipmentId: string): Promise<EquipmentMaster> => {
  const response = await apiClient.get(`/mes/equipment/${equipmentId}`);
  return response.data;
};

export const getEquipmentStatus = async (equipmentId: string): Promise<EquipmentStatus> => {
  const response = await apiClient.get(`/mes/equipment/${equipmentId}/status`);
  return response.data;
};

export const updateEquipmentStatus = async (
  equipmentId: string,
  data: Partial<EquipmentStatus>
): Promise<EquipmentStatus> => {
  const response = await apiClient.post(`/mes/equipment/${equipmentId}/status`, data);
  return response.data;
};

export const getAllEquipmentStatus = async (params?: {
  line_code?: string;
}): Promise<EquipmentCurrentStatus[]> => {
  const response = await apiClient.get('/mes/equipment/status/all', { params });
  return response.data;
};

export const getOEEData = async (params?: {
  line_code?: string;
  equipment_code?: string;
  start_date?: string;
  end_date?: string;
}): Promise<EquipmentOEE[]> => {
  const response = await apiClient.get('/mes/equipment/oee', { params });
  return response.data;
};

export const getOEETrend = async (
  equipmentCode: string,
  days?: number
): Promise<OEETrend[]> => {
  const response = await apiClient.get(`/mes/equipment/oee/${equipmentCode}/trend`, {
    params: { days },
  });
  return response.data;
};

export const getDowntimeEvents = async (params?: {
  page?: number;
  page_size?: number;
  equipment_code?: string;
  line_code?: string;
  status?: string;
  start_date?: string;
  end_date?: string;
}): Promise<DowntimeEvent[]> => {
  const response = await apiClient.get('/mes/equipment/downtime', { params });
  return response.data;
};

export const createDowntimeEvent = async (data: Partial<DowntimeEvent>): Promise<DowntimeEvent> => {
  const response = await apiClient.post('/mes/equipment/downtime', data);
  return response.data;
};

export const updateDowntimeEvent = async (
  eventId: string,
  data: Partial<DowntimeEvent>
): Promise<DowntimeEvent> => {
  const response = await apiClient.patch(`/mes/equipment/downtime/${eventId}`, data);
  return response.data;
};

export const getMaintenanceHistory = async (params?: {
  equipment_code?: string;
  page?: number;
  page_size?: number;
}): Promise<DowntimeEvent[]> => {
  const response = await apiClient.get('/mes/equipment/maintenance', { params });
  return response.data;
};

export const createMaintenanceRecord = async (data: Partial<DowntimeEvent>): Promise<DowntimeEvent> => {
  const response = await apiClient.post('/mes/equipment/maintenance', data);
  return response.data;
};

File: api/schemas/mes/equipment.py
"""
Equipment-related Pydantic schemas
"""
from datetime import datetime, date
from decimal import Decimal
from enum import Enum
from typing import Optional, List, Dict, Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class EquipmentType(str, Enum):
    """Equipment types"""
    LOADER = "loader"
    UNLOADER = "unloader"
    PRINTER = "printer"
    SPI = "spi"
    MOUNTER = "mounter"
    REFLOW = "reflow"
    AOI = "aoi"
    XRAY = "xray"
    WAVE = "wave"
    SELECTIVE = "selective"
    DISPENSER = "dispenser"
    INSERTION = "insertion"
    INSPECTION = "inspection"
    TEST = "test"
    CONVEYOR = "conveyor"
    ROBOT = "robot"
    LASER = "laser"
    COATING = "coating"
    OTHER = "other"


class EquipmentStatusValue(str, Enum):
    """Equipment status values"""
    RUNNING = "running"
    IDLE = "idle"
    SETUP = "setup"
    ALARM = "alarm"
    BREAKDOWN = "breakdown"
    MAINTENANCE = "maintenance"
    OFFLINE = "offline"
    WAITING = "waiting"


class AlarmSeverity(str, Enum):
    """Alarm severity levels"""
    INFO = "info"
    WARNING = "warning"
    ERROR = "error"
    CRITICAL = "critical"
    FATAL = "fatal"


class DowntimeCategory(str, Enum):
    """Downtime categories"""
    BREAKDOWN = "breakdown"
    PLANNED_MAINTENANCE = "planned_maintenance"
    UNPLANNED_MAINTENANCE = "unplanned_maintenance"
    SETUP = "setup"
    CHANGEOVER = "changeover"
    MATERIAL_SHORTAGE = "material_shortage"
    QUALITY_ISSUE = "quality_issue"
    OPERATOR_ABSENCE = "operator_absence"
    WAITING = "waiting"
    OTHER = "other"


# ==================== Production Line ====================

class ProductionLineResponse(BaseModel):
    """Response schema for production line"""
    id: UUID
    tenant_id: UUID
    line_code: str
    line_name: str
    line_type: str
    factory_code: str
    department_code: Optional[str] = None
    capacity_per_shift: Optional[int] = None
    operators_required: int
    status: str
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class LineStatusResponse(BaseModel):
    """Line status with equipment summary"""
    line_code: str
    line_name: str
    line_type: str
    status: str
    equipment_count: int
    running_count: int
    idle_count: int
    down_count: int
    current_oee: Optional[float] = None
    today_production: Optional[int] = None
    today_defects: Optional[int] = None


# ==================== Equipment Master ====================

class EquipmentMasterResponse(BaseModel):
    """Response schema for equipment master"""
    id: UUID
    tenant_id: UUID
    equipment_code: str
    equipment_name: str
    equipment_type: EquipmentType
    equipment_category: Optional[str] = None
    line_code: str
    position_in_line: Optional[int] = None
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    serial_no: Optional[str] = None
    firmware_version: Optional[str] = None
    install_date: Optional[date] = None
    warranty_end_date: Optional[date] = None
    standard_cycle_time_sec: Optional[Decimal] = None
    max_capacity_per_hour: Optional[int] = None
    ip_address: Optional[str] = None
    port: Optional[int] = None
    communication_protocol: Optional[str] = None
    last_pm_date: Optional[date] = None
    next_pm_date: Optional[date] = None
    pm_interval_days: int
    status: str
    attributes: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# ==================== Equipment Status ====================

class EquipmentStatusCreate(BaseModel):
    """Schema for creating equipment status"""
    equipment_code: str = Field(..., max_length=30)
    status: EquipmentStatusValue
    status_reason: Optional[str] = Field(None, max_length=100)
    alarm_code: Optional[str] = Field(None, max_length=30)
    alarm_message: Optional[str] = None
    alarm_severity: Optional[AlarmSeverity] = None
    production_order_no: Optional[str] = Field(None, max_length=20)
    product_code: Optional[str] = Field(None, max_length=30)
    operator_code: Optional[str] = Field(None, max_length=20)


class EquipmentStatusResponse(BaseModel):
    """Response schema for equipment status"""
    id: UUID
    tenant_id: UUID
    equipment_code: str
    status_timestamp: datetime
    status: EquipmentStatusValue
    previous_status: Optional[str] = None
    status_reason: Optional[str] = None
    alarm_code: Optional[str] = None
    alarm_message: Optional[str] = None
    alarm_severity: Optional[AlarmSeverity] = None
    production_order_no: Optional[str] = None
    product_code: Optional[str] = None
    operator_code: Optional[str] = None
    speed_actual: Optional[Decimal] = None
    speed_target: Optional[Decimal] = None
    temperature_actual: Optional[Decimal] = None
    temperature_target: Optional[Decimal] = None
    counters: Optional[Dict[str, Any]] = None
    parameters: Optional[Dict[str, Any]] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class EquipmentCurrentStatus(BaseModel):
    """Current status of equipment with master info"""
    equipment_code: str
    equipment_name: str
    equipment_type: EquipmentType
    line_code: str
    position_in_line: Optional[int] = None
    status: EquipmentStatusValue
    status_since: Optional[datetime] = None
    current_order_no: Optional[str] = None
    product_code: Optional[str] = None
    alarm_code: Optional[str] = None
    alarm_message: Optional[str] = None


# ==================== Equipment OEE ====================

class EquipmentOEEResponse(BaseModel):
    """Response schema for equipment OEE"""
    id: UUID
    tenant_id: UUID
    calculation_date: date
    shift_code: Optional[str] = None
    equipment_code: str
    line_code: str
    planned_time_min: Decimal
    actual_run_time_min: Decimal
    downtime_min: Decimal
    setup_time_min: Decimal
    idle_time_min: Decimal
    ideal_cycle_time_sec: Optional[Decimal] = None
    actual_cycle_time_sec: Optional[Decimal] = None
    total_count: int
    good_count: int
    defect_count: int
    oee: Optional[Decimal] = None
    availability: Optional[float] = None
    performance: Optional[float] = None
    quality: Optional[float] = None
    downtime_breakdown: Optional[Dict[str, Any]] = None
    defect_breakdown: Optional[Dict[str, Any]] = None
    calculated_at: Optional[datetime] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class OEESummary(BaseModel):
    """OEE summary for dashboard"""
    line_code: str
    date: date
    avg_oee: float
    avg_availability: float
    avg_performance: float
    avg_quality: float
    total_downtime_min: float
    equipment_count: int


class OEETrend(BaseModel):
    """OEE trend data point"""
    date: date
    oee: float
    availability: float
    performance: float
    quality: float


# ==================== Downtime Event ====================

class DowntimeEventCreate(BaseModel):
    """Schema for creating downtime event"""
    equipment_code: str = Field(..., max_length=30)
    line_code: str = Field(..., max_length=20)
    start_time: datetime
    downtime_category: DowntimeCategory
    downtime_code: str = Field(..., max_length=20)
    downtime_reason: Optional[str] = None
    alarm_code: Optional[str] = Field(None, max_length=30)
    alarm_message: Optional[str] = None
    production_order_no: Optional[str] = Field(None, max_length=20)
    product_code: Optional[str] = Field(None, max_length=30)
    operator_code: Optional[str] = Field(None, max_length=20)
    reported_by: Optional[str] = Field(None, max_length=20)


class DowntimeEventUpdate(BaseModel):
    """Schema for updating downtime event"""
    end_time: Optional[datetime] = None
    root_cause: Optional[str] = None
    corrective_action: Optional[str] = None
    maintenance_ticket_no: Optional[str] = Field(None, max_length=20)
    impact_qty: Optional[Decimal] = None
    impact_cost: Optional[Decimal] = None
    status: Optional[str] = None
    resolved_by: Optional[str] = Field(None, max_length=20)
    notes: Optional[str] = None


class DowntimeEventResponse(BaseModel):
    """Response schema for downtime event"""
    id: UUID
    tenant_id: UUID
    event_no: str
    equipment_code: str
    line_code: str
    start_time: datetime
    end_time: Optional[datetime] = None
    duration_min: Optional[float] = None
    downtime_category: DowntimeCategory
    downtime_code: str
    downtime_reason: Optional[str] = None
    alarm_code: Optional[str] = None
    alarm_message: Optional[str] = None
    production_order_no: Optional[str] = None
    product_code: Optional[str] = None
    operator_code: Optional[str] = None
    root_cause: Optional[str] = None
    corrective_action: Optional[str] = None
    maintenance_ticket_no: Optional[str] = None
    impact_qty: Optional[Decimal] = None
    impact_cost: Optional[Decimal] = None
    status: str
    resolved_by: Optional[str] = None
    resolved_at: Optional[datetime] = None
    notes: Optional[str] = None
    reported_by: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class DowntimeAnalysis(BaseModel):
    """Downtime analysis summary"""
    line_code: str
    equipment_code: Optional[str] = None
    category: DowntimeCategory
    downtime_code: str
    occurrence_count: int
    total_duration_min: float
    avg_duration_min: float
    percentage: float
